core/src/main/java/org/testcontainers/utility/MountableFile.java

```diff
@@ -0,0 +1,289 @@
+package org.testcontainers.utility;
+
+import com.google.common.base.Charsets;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.apache.commons.lang.SystemUtils;
+import org.jetbrains.annotations.NotNull;
+import org.testcontainers.images.builder.Transferable;
+
+import java.io.*;
+import java.net.URL;
+import java.net.URLDecoder;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import static lombok.AccessLevel.PACKAGE;
+import static org.testcontainers.utility.PathUtils.recursiveDeleteDir;
+
+/**
+ * An abstraction over files and classpath resources aimed at encapsulating all the complexity of generating
+ * a path that the Docker daemon is about to create a volume mount for.
+ */
+@RequiredArgsConstructor(access = PACKAGE)
+@Slf4j
+public class MountableFile implements Transferable {
+
+    private final String path;
+
+    @Getter(lazy = true)
+    private final String resolvedPath = resolvePath();
+
+    @Getter(lazy = true)
+    private final String resolvedPathWithShellEscaping = resolvePathWithShellEscaping();
```

Comment:
FYI about `lazy=true`:

Since `resolvePathWithShellEscaping` is a one-liner, you can put it here without extracting a method. Of course, we can keep a separate method, just saying that it's not a must.